### PR TITLE
util: if `DataDir` is a symbolic link, manually follow it

### LINKED ISSUE #24257: bitcoind and bitcoin-cli does not supports symbolic link anymore
I just pulled latest master at https://github.com/bitcoin/bitcoin/commit/3ace3a17c9bce606cea05192f0da3ac62ac69dda and I cannot run bitcoind or bitcoin-cli without providing a full path for data directory using `-datadir` option. It was working fine couple of days ago (while running https://github.com/bitcoin/bitcoin/commit/807169e10b4a18324356ed6ee4d69587b96a7c70). I suspect it may be related to latest switch to the standard std::filesystem library https://github.com/bitcoin/bitcoin/pull/20744

**Expected behavior**

I use a symbolic link to point my /home/me/.bitcoin to an external hard drive. When running `bitcoind` from command line, or executing `bitcoin-cli` commands, without using the `-datadir` option, it should not terminate or crash unexpectedly. Those commands should be able to follow the symbolic link and be able to read/write into the final folder.

**Actual behavior**

`bitcoind` and `bitcoin-cli` are not able to follow the symbolic link. `bitcoin-qt` runs fine, even when running it from command line. See below commands examples.

```
# Create a symbolic link
$~/btc-things/bitcoin$ ln -s /media/DATA/Bitcoin /home/me/.bitcoin

# Try to run bitcoind - Crashes
$~/btc-things/bitcoin$ ./bitcoind 


************************
EXCEPTION: NSt10filesystem7__cxx1116filesystem_errorE       
filesystem error: cannot create directories: Not a directory [/home/me/.bitcoin]       
bitcoin in AppInit()       

bitcoind: chainparamsbase.cpp:35: const CBaseChainParams& BaseParams(): Assertion `globalChainBaseParams' failed.
Aborted (core dumped)

# Run bitcoin-qt, it works fine, the node runs
$~/btc-things/bitcoin$ ./bitcoin-qt 
qt5ct: using qt5ct plugin

# Try to run a bitcoin-cli command to interact with the running node, does not work
$~/btc-things/bitcoin$ ./bitcoin-cli getblockhash 501726


************************
EXCEPTION: NSt10filesystem7__cxx1116filesystem_errorE       
filesystem error: cannot create directories: Not a directory [/home/me/.bitcoin]       
bitcoin in AppInitRPC()       

# Try the same previous command but give an absolute path for -datadir, works fine
$~/btc-things/bitcoin$ ./bitcoin-cli -datadir=/media/DATA/Bitcoin getblockhash 501726
0000000000000000004b27f9ee7ba33d6f048f684aaeb0eea4befd80f1701126
```

**To reproduce**

Steps are indicated previously.

**System information**

Running `Bitcoin Core version v22.99.0-3ace3a17c9bc`. Self compiled from `master` branch.

I'm using Linux Mint with the following kernel `Linux 5.4.0-96-generic x86_64`.
I'm using an additional hard drive to store the full blockchain (SSD).

I had no prior issues before pulling latest changes. I was previously running at https://github.com/bitcoin/bitcoin/commit/807169e10b4a18324356ed6ee4d69587b96a7c70

### PR DESCRIPTION
**Pull request by nymkappa**
Fixes https://github.com/bitcoin/bitcoin/issues/24257

```
$~/btc-things/bitcoin$ ln -s /media/DATA/Bitcoin /home/me/.bitcoin
$~/btc-things/bitcoin$ ./bitcoind 
************************
EXCEPTION: NSt10filesystem7__cxx1116filesystem_errorE       
filesystem error: cannot create directories: Not a directory [/home/me/.bitcoin]  
```

Another related fix attempt can be found here https://github.com/bitcoin/bitcoin/pull/24266

### CODE REVIEW

File: src/util/system.cpp [hunk 1]
```diff
@@ -443,6 +443,15 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
     } else {
         path = GetDefaultDataDir();
     }
+
+    if (fs::is_symlink(path)) {
+        path = fs::read_symlink(path);
+        if (!fs::is_directory(path)) {
+            path = "";
+            return path;
+        }
```
Comment (hebasto) [inline src/util/system.cpp:452]: What scenarios these lines are required in?

Comment (nymkappa) [reply src/util/system.cpp:452]: A symbolic link may exist but point to a non existing location. For example, if your symbolic link point to an external hard drive, but it is not mounted.

Comment (nymkappa) [reply src/util/system.cpp:452]: For example image

**Comment by nymkappa** [comment]
I've tested the following scenarios:
* Non existing `~/.bitcoin` file/folder - OK (create a `~/.bitcoin` folder)
* Existing `~/.bitcoin` folder  - OK (load the content of `~/.bitcoin` folder)
* Create a new wallet using bitcoin-cli - OK (create `~/.bitcoin/wallets/xxx` folder)
* Moved `~/.bitcoin/wallets/xxx` to `~/.bitcoin/xxx` - OK (I can load `xxx` wallet using bitcoin-cli)
* Create a broken symlink with `ln -s /1234 .bitcoin` - ??? (bitcoind does not start, but seems to crash, see below)
```cpp
bitcoind: fs.cpp:39: fs::path fsbridge::AbsPathJoin(const fs::path&, const fs::path&): Assertion `base.is_absolute()' failed.
Aborted (core dumped)
```

* Create a valid symlink with `ln -s /home/me/Documents .bitcoin` - OK (it read and writes into `/home/me/Documents` properly)

**Comment by hebasto** [comment]
@nymkappa 
> I've tested the following scenarios...

If there are behavior changes comparing to pre-20744 variant, could describe them in the PR description?

**Comment by nymkappa** [comment]
> @nymkappa
> 
> > I've tested the following scenarios...
> 
> If there are behavior changes comparing to pre-20744 variant, could describe them in the PR description?

This is the only behavior change I've noticed from the previously mentioned scenario between [master@04255073bbd2b8ea71ae8a9ff7433be499312758](https://github.com/bitcoin/bitcoin/commit/04255073bbd2b8ea71ae8a9ff7433be499312758) (before #20744 was merged) and my pull request.

It looks like even though `.bitcoin` is a symbolic link (pointing to an invalid location), core does not try to follow it and instead tries to create a directory at that location.

* Create a broken symlink with ln -s /1234 .bitcoin - ??? (bitcoind does not start, but seems to crash, see below)
```
************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::create_directory: File exists: "/home/me/.bitcoin"       
bitcoin in AppInit()       

bitcoind: chainparamsbase.cpp:35: const CBaseChainParams& BaseParams(): Assertion `globalChainBaseParams' failed.
Aborted (core dumped)
```

**Comment by hebasto** [comment]
> This is the only behavior change I've noticed from the previously mentioned scenario between [master@04255073bbd2b8ea71ae8a9ff7433be499312758](https://github.com/bitcoin/bitcoin/commit/04255073bbd2b8ea71ae8a9ff7433be499312758) (before #20744 was merged) and my pull request.

FWIW, #24266 preserves the pre-20744 behavior, and for a broken link `bitcoind` reports an exception with a meaningful message:
```
$ src/bitcoind


************************
EXCEPTION: NSt10filesystem7__cxx1116filesystem_errorE       
filesystem error: cannot create directories: File exists [/home/hebasto/.bitcoin/wallets]       
bitcoin in AppInit()       

```

**Comment by nymkappa** [comment]
> > This is the only behavior change I've noticed from the previously mentioned scenario between [master@04255073bbd2b8ea71ae8a9ff7433be499312758](https://github.com/bitcoin/bitcoin/commit/04255073bbd2b8ea71ae8a9ff7433be499312758) (before #20744 was merged) and my pull request.
> 
> FWIW, #24266 preserves the pre-20744 behavior, and for a broken link `bitcoind` reports an exception with a meaningful message:
> 
> ```
> $ src/bitcoind
> 
> 
> ************************
> EXCEPTION: NSt10filesystem7__cxx1116filesystem_errorE       
> filesystem error: cannot create directories: File exists [/home/hebasto/.bitcoin/wallets]       
> bitcoin in AppInit()       
> ```

Closing this PR as https://github.com/bitcoin/bitcoin/pull/24266 provides a better fix. Thank you for your time hebasto 👍🏻

**Pull request closed without merging**